pkg/blockstorage/zone/sanitize_zones.go

```diff
@@ -0,0 +1,139 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Parts of this code are derived from https://github.com/agnivade/levenshtein
+// which carries the following license
+
+// The MIT License (MIT)
+// Copyright (c) 2015 Agniva De Sarker
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package zone
+
+import (
+	"sort"
+	"unicode/utf8"
+
+	"github.com/kanisterio/kanister/pkg/field"
+	"github.com/kanisterio/kanister/pkg/log"
+)
+
+func sanitizeAvailableZones(availableZones map[string]struct{}, validZoneNames []string) map[string]struct{} {
+	sanitizedZones := map[string]struct{}{}
+	for zone := range availableZones {
+		if isZoneNameValid(zone, validZoneNames) {
+			sanitizedZones[zone] = struct{}{}
+		} else {
+			closestMatch := levenshteinMatch(zone, validZoneNames)
+			log.Debug().Print("Exact match not found for available zone, using closest match",
+				field.M{"availableZone": zone, "closestMatch": closestMatch})
+			sanitizedZones[closestMatch] = struct{}{}
+		}
+	}
+	return sanitizedZones
+}
+
+func isZoneNameValid(zone string, validZones []string) bool {
+	for _, z := range validZones {
+		if zone == z {
+			return true
+		}
+	}
+	return false
+}
+
+func levenshteinMatch(input string, options []string) string {
+	sort.Slice(options, func(i, j int) bool {
+		return computeDistance(input, options[i]) < computeDistance(input, options[j])
+	})
+	return options[0]
+}
+
+// ComputeDistance computes the levenshtein distance between the two
```

Comment:
[nit] please separate the copied code as much as possible I would put this whole method in another file.